### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive with core [require-atomic-updates] conflict

### ISSUE THREAD

**Issue by brokentone** [typescript-eslint/typescript-eslint#11857]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBBA2IDuAxArmYUCW4YF4YBDCAT0xgAoBKAgPhgG8AoGGAJwFMo12wYARADci8NJwEBuZgF9mzeNxgALbAHNlndgGVQAB04A1UeIBccKO2xg1MAD4wMAE04Aza5yfTmJcsBiuGFi4-FCc0OiYOOA0TKww2K5UqhpauiAGxmKcBPiEzm4eTrQsbGygkLBcEGjwsIRESETYsAjIkcEx1NJlCUmUKZo6%2BkYmOXn5YC7uYJ4l8b2DaSNZ4gQc4bVQPWVybHIyQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEI0QR1gEtpEBaAQwBcB7AW0IGNTYAHAEypXSkWmmuiQANOGyQAApQCeLFA2iEWlUiniEAdpQD066s3XrEDFMnLQppBtXVtClQte4BtSL36ChmKAwAWRgNYAKjKIAAoktgycqBiUcIhgAL5gALqiiSCJQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const slowFunction = async () => {
  return "value";
}

let higherScopeValue: string | undefined;

async function testFunction() {
  if (higherScopeValue === undefined) {
    const result = await slowFunction();
    if (higherScopeValue === undefined) {
      higherScopeValue = result;
    }
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "require-atomic-updates": "error",
    "@typescript-eslint/no-unnecessary-condition": ["error", { "checkTypePredicates": true } ]
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect this secondary check does not error, as it is necessary to avoid the race condition which is introduced in the await between the 2 conditions.

### Actual Result

Unnecessary conditional, comparison is always true, since `undefined === undefined` is true. 10:9 - 10:39

### Additional Info

I understand the current limitations of this rule regarding https://typescript-eslint.io/rules/no-unnecessary-condition#values-modified-within-function-calls but I'm fairly confident that this is not a limitation to this specific false positive. My rationale for this is that doesn't depend on a specific value of the parent scope, only that any check of a value that is outside the scope that contains an await is no longer valid after the resuming of the await. Also being that the `require-atomic-updates` rule is able to find this exact issue.

**Comment by kirkwaiblinger** [comment]
Hi @brokentone!

TS doesn't treat `await`s specially when it comes to narrowing... If you hover over `higherScopeValue` at the site of the lint report ([playground](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBBA2IDuAxArmYUCW4YF4YBDCAT0xgAoBKAgPhgG8AoGGAJwFMo12wYARADci8NJwEBuZgF9mzeNxgALbAHNlndgGVQAB04A1UeIBccKO2xg1MAD4wMAE04Aza5yfTmJcsBiuGFi4-FCc0OiYOOA0TKww2K5UqhpauiAGxmKcBPiEzm4eTrQsbGygkLBcEGjwsIRESETYsAjIkcEx1NJlCUmUKZo6%2BkYmOXn5YC7uYJ4lzAD0C71sAHoA-Csw8b2DaSNZ4gQc4bVQPWVybHIyQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEI0QR1gEtpEBaAQwBcB7AW0IGNTYAHAEypXSkWmmuiQANOGyQAApQCeLFA2iEWlUiniEAdpQD066s3XrEDFMnLQppBtXVtClQte4BtSL36ChmKAwAWRgNYAKjKIAAoktgycqBiUcIhgAL5gALqiiSCJQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)), you'll see the type is just `undefined`. That's all the rules goes off of, and accordingly this is a classic limitation of the rule.

You can find discussion of this on the TS side at, e.g. https://github.com/microsoft/TypeScript/issues/31429, https://github.com/microsoft/TypeScript/issues/27900

**Comment by brokentone** [comment]
Hi @kirkwaiblinger -- Thanks for the really speedy response. I see what you're saying -- it's less of *how* it's narrowed, or the relationship between the value of the variables referenced in various moments, and it's simply *that* it was narrowed which then triggers this rule.

It's unfortunate this can't be resolved in that this rule then might be unintentionally steering folks to make updates which are then creating race conditions (in my exact use case, but certainly not every case). But I see clearly that this is a limitation of the current maturity of the typescript engine itself not this ruleset. And, circumventing that here, building negative cases on top of this positive case, would cause major complexity sprawl.

Thanks to you and the team for all your work, really helpful tool for me any my teams!

**Comment by kirkwaiblinger** [comment]
> It's unfortunate this can't be resolved in that this rule then might be unintentionally steering folks to make updates which are then creating race conditions (in my exact use case, but certainly not every case). But I see clearly that this is a limitation of the current maturity of the typescript engine itself not this ruleset. And, circumventing that here, building negative cases on top of this positive case, would cause major complexity sprawl.

Yeah, like you say, it's a question of whether to work with or against the grain of TS.

TS also lets you do something totally unsafe because of this behavior ([playground](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBBA2IDuAxArmYUCW4YF4YAKASgID4YwBTJGABQCcQBbbCagHgDcRsATckUbUIIeN2oUYAbwBQMONSgAVbC2og0UIqWkyYIsROp6AvgBoYARgAM9kgG45Zp3LnxlMABbYA5t7UjADKoAAO1ABqAIbwaNQAXHBQjNhgfjAAPjAY-NQAZmnU-M5y0RAAnpgw%2BRhYuGAwUKJQ6Jg44ACqkNH5pmTyitj5xL4BQaEgETFxUgCE%2BIS5BUX8AwqKMNFI0diwCMht9eCkzpswoJDi1AB0iH5EY4Eh4VGx8TdQIJ1hEYwAwuV%2Bk4YAB6UEwMIgCAQbAAI08hgwOA0F0Y5W85w2Zhc7g4UAAkmBmoxuLFdGR8JRBj5-M9JtN3lIlmA8oUaCUXFY7LY3GVKtVau0Gk0WkcOmBgr1%2BrINsNRnSJq8ZvEYAsWWzVutzttdvtEKg6hLThtFOCYABHNB7KQ0YCiCDRRgVOammDyx6Kl5TN6zNWLHKslYc7XnC7gYy3e6e8bexmzT7fX5BQEcUgg82OvpunGKHFmIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEI0QR1gEtpEBaAQwBcB7AW0IGNTYAHAEypXSkWmmuiQANOGyQAApQCeLFA2iEWlUiniEAdpQD066s3XrEDFMnLQppBtXVtClQte4BtSL36ChmKAwAWRgNYAKjKIAAoktgycqBiUcIhgAL5gALqiiSCJQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
setInterval(() => {
  higherScopeValue = undefined;
}, 100)

async function testFunctionUnsafe() {
  if (higherScopeValue !== undefined) {
    await slowFunction();
    console.log(higherScopeValue.toUpperCase()); // possible runtime crash    
  }
}
``` 

That's an intentional design decision and we'd be working against the grain to indicate reports otherwise. Usually my approach with TS would be to a) use `const` as much as possible and b) use explicit getter function calls (`getHigherScopeValue()`) to query state that may change. 

For example:

```ts
async function testFunction() {
   const oldValue = getHigherScopeValue();
   if (oldValue === undefined) {
      const result = await slowFunction();
      const newValue = getHigherScopeValue();
      if (newValue === undefined) {
          setHigherScopeValue(result);
      }
   }
}
```

Many possible to go about this, but yeah it's an area where you might end up writing "unnecessary" ugly JS for the sake of satisfying the type checker and TS's very limited mutability analysis.

> Thanks to you and the team for all your work, really helpful tool for me any my teams!

❤️
